### PR TITLE
Support single sign-on (SSO)

### LINKED ISSUE #86: Support OpenID Connect / Social Login
We need a way to make it easy to configure an Open Liberty deployment to take advantage of OpenID Connect / Social Login capabilities.

Some of this work will be at the container level via server configuration (with environment variables place holders), but the more complex part of this work will be at the Operator level to fetch the relevant data and populate the snippets (via env vars).   Also need to consider doing file mounts, if the data is sensitive (e.g. credentials).  

This probably needs to work with a set of OpenID Connect Providers (Keycloak, Open Liberty OP, etc).

### PR DESCRIPTION
**Pull request by leochr**
**What this PR does / why we need it?**:

- Add support for single sign-on (sso)

Todos:
Rebase on top of the refactored code
create a separate PR to update documentation and changelog

**Which issue(s) this PR fixes**:

Fixes #86

### CODE REVIEW

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -189,3 +197,143 @@ func ConfigureServiceability(pts *corev1.PodTemplateSpec, la *openlibertyv1beta1
...
+	case string:
+		return v.(string)
+	case bool:
+		return strconv.FormatBool(v.(bool))
+	default:
+		return ""
+	}
+}
+
+func getSsoEnv(loginID string, envSuffix string, value interface{}) *corev1.EnvVar {
```
Comment (navidsh) [inline pkg/utils/utils.go]: To be consistent with other parts of the code, what do you think about `getSSOEnv?

File: pkg/controller/openliberty/openlibertyapplication_controller.go [hunk 2]
```diff
@@ -375,6 +398,16 @@ func (r *ReconcileOpenLiberty) Reconcile(request reconcile.Request) (reconcile.R
 			autils.CustomizePodSpec(&statefulSet.Spec.Template, instance)
 			autils.CustomizePersistence(statefulSet, instance)
 			lutils.CustomizeLibertyEnv(&statefulSet.Spec.Template, instance)
+			if instance.GetSSO() != nil {
+				secretName := instance.GetName() + ssoSecretNameSuffix
+				ssoSecret := &corev1.Secret{}
+				err := r.GetClient().Get(context.TODO(), types.NamespacedName{Name: secretName, Namespace: instance.GetNamespace()}, ssoSecret)
+				if err != nil {
+					reqLogger.Error(nil, "Secret for Single sign-on (SSO) was not found. Create a secret named '"+secretName+"' in namespace '"+instance.GetNamespace()+"' with the credentials for the login providers you selected in application image.")
```
[the remark this answers is not in the document]
Comment (navidsh) [reply pkg/controller/openliberty/openlibertyapplication_controller.go]: e.g.
```go
errors.Wrapf(err, "unable to fetch credential %q from secret %q", cred, sel.Name)
```

[the remark this answers is not in the document]
Comment (navidsh) [reply pkg/controller/openliberty/openlibertyapplication_controller.go]: `%q` adds quotes around a value i.e. https://play.golang.org/p/VT47P5ix1RP

File: pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go [hunk 3]
```diff
@@ -468,6 +533,167 @@ func (m *OpenLibertyApplicationMonitoring) GetEndpoints() []prometheusv1.Endpoin
...
+// GetUserApiType returns userApiType
+func (o *OAuth2Client) GetUserApiType() string {
+	return o.UserApiType
+}
+
+// GetUserApi returns userApi
+func (o *OAuth2Client) GetUserApi() string {
+	return o.UserApi
+}
+
```
Comment (arturdzm) [inline pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go]: I know its trying to be consistent with BaseComponent interface impl. But having getter on simple structs (not interface) is redundant and is anti-pattern in Go

File: pkg/controller/openliberty/openlibertyapplication_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (leochr) [reply pkg/controller/openliberty/openlibertyapplication_controller.go]: Good suggestion. Addressed.

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -122,6 +125,11 @@ func findEnvVar(name string, envList []corev1.EnvVar) (*corev1.EnvVar, bool) {
 	return nil, false
 }
 
+// normalizeEnvVariableName Normalize env variable name. Replaces '-' and '.' characters with '_' and converts all characters to upper case
+func normalizeEnvVariableName(name string) string {
```
[the remark this answers is not in the document]
Comment (leochr) [reply pkg/utils/utils.go]: The function from RC is not exported. In next release we can export and then use it in OL.

File: pkg/utils/utils.go (hunk 1 above)
Comment (leochr) [reply pkg/utils/utils.go]: The naming guidelines suggested capitalizing only the first letter of abbreviations that are in the middle. It does look weird. Changed name to `createEnvVarSSO` :)

File: pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go (hunk 3 above)
Comment (leochr) [reply pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go]: Good point. Addressed.

File: pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go [hunk 5]
```diff
@@ -174,6 +175,65 @@ type OpenLibertyApplicationList struct {
...
+	AccessTokenRequired     *bool  `json:"accessTokenRequired,omitempty"`
+	AccessTokenSupported    *bool  `json:"accessTokenSupported,omitempty"`
+	UserApiType             string `json:"userApiType,omitempty"`
+	UserApi                 string `json:"userApi,omitempty"`
+}
+
+// GithubLogin represents configuration for social login using GitHub.
+// +k8s:openapi-gen=true
+type GithubLogin struct {
+	Hostname string `json:"hostname,omitempty"`
```
[the remark this answers is not in the document]
Comment (leochr) [reply pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go]: Hostname is one word. It'll be ideal to use `HostnameVerificationEnabled`, but Liberty config uses `hostNameVerificationEnabled`. Split between keeping the name same as/aligned with Liberty or being consistent on the operator side. Let's chat tomorrow and decide.

File: deploy/releases/daily/openliberty-app-operator.yaml [hunk 6]
```diff
@@ -55,18 +53,6 @@ rules:
...
-  resources:
-  - pods
-  verbs:
-  - get
-- apiGroups:
-  - apps
-  resources:
-  - replicasets
-  verbs:
-  - get
```
Comment (leochr) [inline deploy/releases/daily/openliberty-app-operator.yaml]: @arturdzm @navidsh This portion was removed from role and cluster role. Just want to confirm that it's not needed.

Comment (navidsh) [reply deploy/releases/daily/openliberty-app-operator.yaml]: I don't think we do anything specific to `replicasets`. Not sure about `pod`. Isn't that needed for trace/dump?

File: pkg/utils/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (navidsh) [reply pkg/utils/utils.go]: Sounds good.

File: pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (navidsh) [reply pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go]: As discussed, let's keep it in-sync with Open Liberty server configuration naming convention.

**Pull request merged**